basis/math/matrices/matrices.factor

```diff
@@ -197,77 +297,316 @@ ERROR: negative-power-matrix m n ;
+    [ rest ] [ [ first ] [ third ] bi 2array ] [ 1 head* ] tri 3array
+    ! find determinants
+    [ (2determinant) ] map
+    ! negate odd elements of a b c and multiply by the new determinants
+    { a b c } t alternating-sign v*
+    ! sum the resulting sequence
+    sum ;
+
+DEFER: (ndeterminant)
+: make-determinants ( n matrix -- seq )
+    <repetition> [
+        cols-except [ length ] keep (ndeterminant) ! recurses here
+    ] map-index ;
+
+DEFER: (determinant)
+! generalized to 4 and higher
+: (ndeterminant) ( n matrix -- ndet )
+    ! TODO? recurse for n < 3
+    over 4 < [ (determinant) ] [
+        [ nip first t alternating-sign ] [ rest make-determinants ] 2bi
+        v* sum
+    ] if ;
+
+! switches on dimensions only
+: (determinant) ( n matrix -- determinant )
+    over {
+        { 1 [ nip (1determinant) ] }
+        { 2 [ nip (2determinant) ] }
+        { 3 [ nip (3determinant) ] }
+        [ drop (ndeterminant) ]
+    } case ;
+PRIVATE>
+
+GENERIC: determinant ( matrix -- determinant )
+M: zero-square-matrix determinant
+    drop 0 ;
+
+M: square-matrix determinant
+    [ length ] keep (determinant) ;
+
+! determinant is undefined for m =/= n, unlike inverse
+M: matrix determinant
+    dim first2 non-square-determinant ;
+
+: 1/det ( matrix -- 1/det )
+    determinant recip ; inline
+
+! -----------------------------------------------------
+! inverse operations and implementations follow
+ALIAS: additive-inverse mneg
+
+! per element, find the determinant of all other elements except the element's row / col
+! https://www.mathsisfun.com/algebra/matrix-inverse-minors-cofactors-adjugate.html
+: >minors ( matrix -- matrix' )
+    matrix-except-all [ [ determinant ] map ] map ;
+
+! alternately invert values of the matrix (see alternating-sign)
+: >cofactors ( matrix -- matrix' )
+    [ even? alternating-sign ] map-index ;
+
+! multiply a matrix by the inverse of its determinant
+: m*1/det ( matrix -- matrix' )
+    [ 1/det ] keep n*m ; inline
+
+! inverse implementation
+<PRIVATE
+! https://www.mathsisfun.com/algebra/matrix-inverse-minors-cofactors-adjugate.html
+: (square-inverse) ( square-matrix -- inverted )
+    ! inverse of the determinant of the input matrix
+    [ 1/det ]
+    ! adjugate of the cofactors of the matrix of minors
+    [ >minors >cofactors transpose ]
+    ! adjugate * 1/det
+    bi n*m ;
+
+! TODO
+: (left-inverse) ( matrix -- left-invert )   ;
+: (right-inverse) ( matrix -- right-invert ) ;
+
+! TODO update this when rank works properly
+! only defined for rank(A) = rows(A) OR rank(A) = cols(M)
+! https://en.wikipedia.org/wiki/Invertible_matrix
+: (specialized-inverse) ( rect-matrix -- inverted )
+    dup [ rank ] [ dim ] bi [ = ] with map {
+        { { t f } [ (left-inverse) ] }
+        { { f t } [ (right-inverse) ] }
+        [ no-case ]
+    } case ;
+PRIVATE>
+
+! A^-1
+GENERIC: multiplicative-inverse ( matrix -- inverse )
+M: zero-square-matrix multiplicative-inverse
```

Comment:
Perhaps you can add methods to the recip generic instead of defining a new one?

Reply:
of course, it makes sense to do that -- now in `extra` :)